pages/docs/manual/latest/polymorphic-variant.mdx

```diff
@@ -0,0 +1,273 @@
+
+Check the output in these examples:
+
+
+<CodeTab labels={["ReScript", "JS Output"]}>
+
+```res example
+let capitalized = #Hello
+let lowercased = #goodbye
+
+let err = #error("oops!")
+
+let num = #\"1"
+```
+
+```js
+var capitalized = "Hello";
+
+var lowercased = "goodbye";
+
+var err = {
+  NAME: "error",
+  VAL: "oops!"
+};
+
+var num = "1";
+```
+
+</CodeTab>
+
+
+**Note:** Poly variants play an important role for binding to JS functions in existing JavaScript. Check out the [Bind to JS Function page](bind-to-js-function#constrain-arguments-better) to learn more.
+
+
+### Bind to String Enums
+
+Let's assume we have a TypeScript module that expresses following (stringly typed) enum export:
+
+```js
+// direction.js
+enum Direction {
+  Up = "UP",
+  Down = "DOWN",
+  Left = "LEFT",
+  Right = "RIGHT",
+}
+
+export const myDirection = Direction.Up
+```
+
+For this particular example, we can use poly variants to design the type for the imported `myDirection` value:
+
+
+<CodeTab labels={["ReScript", "JS Output"]}>
+
+```res
+type direction = [ #UP | #DOWN | #LEFT | #RIGHT ]
+@bs.module("./direction.js") external myDirection: direction = "myDirection"
+```
+
+```js
+var DirectionJs = require("./direction.js");
+
+var myDirection = DirectionJs.myDirection;
+```
+
+</CodeTab>
+
+Since we were using poly variants, the JS Output is practically zero-cost and doesn't add any extra code!
+
+## Lower / Upper Bound Constraints
+
+There are a few different ways to define constraints on a poly variant type, such as `[>`, `[<` and `[`. Some of them were briefly mentioned before, so in this section we will quickly explain what this syntax is about.
+
+**Note:** We added this info for educational purposes. In most cases you will not want to use any of this stuff, since it makes your APIs pretty unreadable / hard to use.
+
+### Exact (`[`)
+
+This is the simplest poly variant definition, and also the most practical one. Like a common variant type, this one defines an exact set of constructors.
+
+```res
+type rgb = [ #Red | #Green | #Blue ]
+
+let color: rgb = #Green
+```
+
+In the example above, `color` will only allow one of the three constructors that are defined in the `rgb` type. This is usually the way how poly variants should be defined.
+
+In case you want to define a type that is extensible in polymorphic ways (or in other words, subtyping allowed sets of constructors), you'll need to use the lower / upper bound syntax. 
+
+### Lower Bound (`[>`)
+
+A lower bound defines the minimum set of constructors a poly variant type is aware of. It is also considered an "open poly variant type", because it doesn't restrict any additional values.
+
+Here is an example on how to make a minimum set of `basicBlueTones` extensible for a new `colors` type:
+
+```res
+type basicBlueTone<'a> = [> #Blue | #DeepBlue | #Azuro ] as 'a
```

Comment:
[Azure](https://en.wikipedia.org/wiki/Azure_(color))?

Reply:
Would an explanation of the `as 'a`  be needed here?

Reply:
yeah it would, but not sure how to explain it in words without any concrete example (without being overly abstract)

Reply:
The `as 'a` makes a lot more sense once we understand "type quantification," which I think is pretty foreign/arcane to a lot of people. But it might be helpful to allude to some of the **other** ways that types are quantified, and how the `as` keyword plays a role in determining how "type holes" are "filled."

Reply:
We will have a dedicated section about polymorphism in an advanced section... i think i will go with a note that does some oversimplification for the case at hand, and later on point to more advanced resources when they are ready.